V2 mirror support

RichardScothern opened the pull request:
The v2 registry will act as a pull-through cache, and needs to be
handled differently by the client to the v1 registry mirror.

See https://github.com/docker/distribution/issues/459 for details

Configuration

Only one v2 registry can be configured as a mirror.  Acceptable configurations
in this chanage are: 0...n v1 mirrors or 1 v2 mirror.  A mixture of v1 and v2
mirrors is considered an error.

Pull

If a v2 mirror is configured, all pulls are redirected to that mirror.  The
mirror will serve the content locally or attempt a pull from the upstream mirror,
cache it locally, and then serve to the client.

Push

If an image is tagged to a mirror, it will be pushed to the mirror and be
stored locally there.  Otherwise, images are pushed to the hub.  This is 
unchanged behavior.

This change should not affect the v1 codepath.  

closes docker/distribution#84
Connects to docker/distribution#19,  docker/distribution#459

rscothern committed:
Registry v2 mirror support.

The v2 registry will act as a pull-through cache, and needs to be
handled differently by the client to the v1 registry mirror.

See docker/distribution#459 for details

Configuration

Only one v2 registry can be configured as a mirror. Acceptable configurations
in this chanage are: 0...n v1 mirrors or 1 v2 mirror. A mixture of v1 and v2
mirrors is considered an error.

Pull

If a v2 mirror is configured, all pulls are redirected to that mirror. The
mirror will serve the content locally or attempt a pull from the upstream mirror,
cache it locally, and then serve to the client.

Push

If an image is tagged to a mirror, it will be pushed to the mirror and be
stored locally there. Otherwise, images are pushed to the hub. This is
unchanged behavior.

Signed-off-by: Richard Scothern <richard.scothern@gmail.com>

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "v2-mirror" git@github.com:RichardScothern/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

stevenschlansker commented:
I strongly disagree with "Only one v2 registry can be configured as a mirror."
Having multiple fallback mirrors is incredibly useful, as it allows much easier HA without complicated server configuration.  It also allows multi-datacenter availability of registries which is even harder to configure on the server side.

Compare configuring your daemon as `--add-mirror mycorp-mirror-uswest.mycorp.com --add-mirror mycorp-mirror-useast.mycorp.com` versus attempting to create a highly available server side mirror spanning the US west and east coasts.

This was originally filed in https://github.com/docker/docker/pull/9161

RichardScothern commented:
@stevenschlansker .  I appreciate your point, but you can get the same effect with a load balancer pointing to your two mirror instances without too much complication.  

Putting load balancing code in the client which deals with all edge cases would complicate this feature and is outside the scope of this change.

dmp42 commented:
@RichardScothern agreed.

Thanks @stevenschlansker 
Like I said in the other ticket, it's certainly a great idea for the future, and I think it can be implemented as a compatible extension later on, on top of this.

It's still going to make the code more complex, and needs to be though through (specifically with regard to delegated delivery of layers and failure handling at that level).

jfrazelle commented:
is this  dependent on #13375

stevvooe commented:
This is looking good. We may need a few tests.

dmcgowan commented on graph/pull.go:
```diff
@@ -102,6 +113,94 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
+		return v2MirrorEndpoint, v2MirrorRepoInfo, nil
+	}
+	if v2MirrorEndpoint != nil && v1MirrorCount > 0 {
+		lastErr = fmt.Errorf("v1 and v2 mirrors configured")
```
Why is this an error?

RichardScothern replied:
Since the behavior's are so different, I don't see how mixing v1 and v2 mirrors would work well.  We can avoid this by having backwards compatibility (just v1 mirrors) or new behavior (a v2 mirror)

dmcgowan commented:
I have some concerns about the change of behavior from v1 mirrors.  The v1 mirrors were only intended to be used for mirroring the official index. While this is an absolutely horrendous feature and requirement, it does have some relevance here. Without sending the hostname as part of the repo name, the mirror will not know which registry to fetch the content from. The assumption based on the current flag is that the content would come from the official index. If you try and send the hostname as part of the reponame, I believe the current path regexs will choke.

How would the mirror behave between these two requestss
```
docker pull myregistry.com/dmcgowan/myapp
docker pull dmcgowan/myapp
```

Are there any instructions on how to test this feature?

tiborvass commented on graph/pull.go:
```diff
@@ -102,6 +113,94 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
+		registry.DockerHeaders(imagePullConfig.MetaHeaders)...,
+	)
+	client := registry.HTTPClient(tr)
+	mirrorSession, err := registry.NewSession(client, imagePullConfig.AuthConfig, mirrorEndpoint)
```
@RichardScothern did you verify whether this will send credentials to the mirror?

dmcgowan replied:
This will send credentials, this needs to clear out the AuthConfig

RichardScothern commented:
@dmcgowan : the behavior does change between registry versions.  Hence the fallback to v1 behavior with v1 mirrors and the inability to mix v1 and v2.  In your scenario the registry would have to deal with a hostname and implement behavior on checking upstream for myregistry.com

Testing this feature is a bit tricky since we don't have the implementation in the v2 registry yet.  I was testing with a pre-loaded v2 registry as a mirror.

dmcgowan commented:
@RichardScothern Created a PR against your fork to make it official only. If you don't have time to merge it I will carry the PR. I cannot sign off on this PR without either limiting it to official only (ugly but current behavior) or making this is a much larger change to support a the full namespace format in the URL generation and support sending the canonical name as the remote name. The first is less controversial and more likely to get merged. The latter we WILL do, but will require more discussion and sign off for the namespace proposals.

dmcgowan committed:
Allow mirroring only for the official index

Strip authconfig from session to keep credentials from being sent to the mirror.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

RichardScothern commented:
Closing - this won't make 1.7.0.

This feature will be far more useful with full namespace support.

RichardScothern commented:
Re-opened.  We would like to have feature parity with the v1 registry in 1.7

tiborvass commented:
design +1, this is only for the official index, and does not send hub creds to the mirror.
It is incompatible with v1 mirrors, and there can only be one v2 mirror set, at least for now in this PR. Rationale: if people want to do some fancy things, they can set up load balancers and set the v2 mirror to that, basically eliminating the need to complexify the client.

tiborvass commented:
Code LGTM

jfrazelle commented on graph/pull.go:
```diff
@@ -102,6 +116,93 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
+	v1MirrorCount := 0
+	var v2MirrorEndpoint *registry.Endpoint
+	var v2MirrorRepoInfo *registry.RepositoryInfo
+	var lastErr error
```
but why?

tiborvass replied:
why what

jfrazelle replied:
i just dont really like it... but i guess i see the point

calavera commented:
it would be cool to have some tests around mirror configuration, because looking at the code I don't really know if it works :smile:

calavera commented on graph/pull.go:
```diff
@@ -55,6 +55,25 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
+	if repoInfo.Index.Official {
+		v2mirrorEndpoint, v2mirrorRepoInfo, err := configureV2Mirror(repoInfo, s.registryService)
+		if err != nil {
+			logrus.Errorf("Error configuring mirrors: %s", err)
```
this log might not be necessary if we're returning the error.

rscothern committed:
Restore the stripped registry version number

Signed-off-by: Richard Scothern <richard.scothern@gmail.com>

tiborvass commented on graph/pull.go:
```diff
@@ -181,6 +282,8 @@ func (s *TagStore) pullRepository(r *registry.Session, out io.Writer, repoInfo *
 			var isDownloaded bool
 			for _, ep := range repoInfo.Index.Mirrors {
+				// Ensure endpoint is v1
+				ep = ep + "/v1/"
```
`ep += "v1/"` there is already a trailing slash at the end of `ep`.

tiborvass commented on graph/pull.go:
```diff
@@ -55,6 +55,25 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
+		}
+
+		if v2mirrorEndpoint != nil {
+			logrus.Debugf("Attempting pull from v2 mirror: %s", v2mirrorEndpoint.URL)
```
`Attempting to pull`

tiborvass commented on graph/pull.go:
```diff
@@ -102,6 +116,93 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
+
+func configureV2Mirror(repoInfo *registry.RepositoryInfo, s *registry.Service) (*registry.Endpoint, *registry.RepositoryInfo, error) {
+	mirrors := repoInfo.Index.Mirrors
+	if len(mirrors) == 0 && !repoInfo.Index.Official {
```
@RichardScothern this is weird, you only call `configureV2Mirror` if `repoInfo.Index.Official` is true, so this block can be ommitted. Cf https://github.com/docker/docker/pull/13374/files#diff-7cbe4277b9a08ded689d34eaae37655eR64

RichardScothern replied:
When the restriction on the official index is lifted the logic will still hold, so I'd prefer to leave it in.

tiborvass replied:
@RichardScothern Provided the restriction is lifted, why would this return mirrors to the official index when the index was marked as non-official?

dmcgowan commented:
LGTM

rscothern committed:
Review feedback:
    - Match verbiage with other output
    - Remove dead code and clearer flow

Signed-off-by: Richard Scothern <richard.scothern@gmail.com>

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "v2-mirror" git@github.com:RichardScothern/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

calavera commented:
@RichardScothern no signature in that latest commit :godmode:

Please, add some tests :pray:

tiborvass commented on graph/pull.go:
```diff
@@ -102,6 +116,93 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
+
+	if v1MirrorCount == len(mirrors) {
+		// OK, but mirrors are v1
+		return nil, nil, nil
```
@RichardScothern this is the exit when there are no mirrors at all, we should probably just have a very early return if `len(mirrors) == 0` at the beginning of this function, otherwise this is confusing because it just so happens that v1MirrorCount would also be 0.

RichardScothern replied:
yep

jfrazelle commented:
LGtM
